pkg/registry/core/pod/storage/eviction.go

```diff
@@ -307,12 +308,20 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
 	return &metav1.Status{Status: metav1.StatusSuccess}, nil
 }
 
-func addConditionAndDeletePod(r *EvictionREST, ctx context.Context, name string, validation rest.ValidateObjectFunc, options *metav1.DeleteOptions) error {
-	if feature.DefaultFeatureGate.Enabled(features.PodDisruptionConditions) {
-		pod, err := getPod(r, ctx, name)
-		if err != nil {
-			return err
+func addConditionAndDeletePod(r *EvictionREST, ctx context.Context, name string, pod *api.Pod, deleteOptions *metav1.DeleteOptions) error {
+	var err error
+	deleteOptionsCopy := deleteOptions.DeepCopy()
+
+	if !dryrun.IsDryRun(deleteOptionsCopy.DryRun) && feature.DefaultFeatureGate.Enabled(features.PodDisruptionConditions) {
+		podCopy := pod.DeepCopy()
+		// morph the pod into the pod requested by preconditions and let Update fail in case they do not pass
+		if !resourceVersionIsUnset(deleteOptionsCopy) {
+			podCopy.ResourceVersion = *deleteOptionsCopy.Preconditions.ResourceVersion
 		}
+		if deleteOptionsCopy.Preconditions != nil && deleteOptionsCopy.Preconditions.UID != nil {
+			podCopy.UID = *deleteOptionsCopy.Preconditions.UID
+		}
```

Comment:
oh! I just got oriented to what's happening here, and there's a much more elegant approach that gets rid of the race between `getPod` and `Update` entirely

Update already gets from etcd internally, then hands a copy of the old pod to the UpdatedObjectInfo getter and passes it through the transformer chain

Rather than plumbing `pod` to addConditionAndDeletePod`, update our transformer chain to transform the pod coming from etcd, adding in the resourceVersion and UID precondition if applicable, along with our conditions:

```go
setPreconditions := func(_ context.Context, _, oldObj runtime.Object) (runtime.Object, error) {
  oldPodCopy := oldObj.(*api.Pod).DeepCopy()
  if options.Preconditions != nil {
    if uid := options.Preconditions.UID; uid != nil && len(*uid) > 0 {
      oldPodCopy.UID = *uid
    }
    if rv := options.Preconditions.ResourceVersion; rv != nil && len(*rv) > 0 {
      oldPodCopy.ResourceVersion = *rv
    }
  }
  return oldPodCopy, nil
}
conditionAppender := ...
conditionedPod := rest.DefaultUpdatedObjectInfo(nil, setPreconditions, conditionAppender)
```

if the precondition uid or resourceVersion conflicts with what's already in etcd, Update will return a conflict appropriately

if there was no precondition uid or resourceVersion, Update will handle retrying internally as needed

then continue with what you have here, updating the resourceVersion in `options` with the resulting resourceVersion from our Update call

Reply:
Nice, this makes more sense. Thanks a lot for digging into this. I have used your code snippet and added some  comments. Can you please let me know if the comments make sense or if I should make them more verbose? Eg. mentioning the logic of the Update function?